docs(rfc): add declarative configuration rfc and rfc template (#328)
getoutreach/devbase: rfcs/328-declarative-configuration.md

  - [Testing Framework](#testing-framework)
  - [Configuration Examples](#configuration-examples)
- [Testing and Release Plan](#testing-and-release-plan)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

This proposal replaces environment variable based configuration being used in Make+Magefiles in favor of using declarative yaml based configuration.

## Motivation

As we've been writing `devbase`, and other tooling that uses it, we've identified that the current practice of environment variable based configuration as well as opinionated and unchangeable defaults is not flexible enough for using this tooling both outside of stencil-base (where it is primarily used today) and giving our users the flexibility to easily make minor behaviour changes. This has, ultimately created two pain points:

 * Documentation is hard to write and not easily discoverable
 * Finding configuration consumption, and configurable areas is difficult
 * Using `devbase` outside of `stencil-base` projects is not easy (must use sub-modules, and even then it's hard to "plug and play") 

### Goals

>  - All configuration should be written as go structs (as we intend to only write new targets in), and have godoc compatible documentation on each exported field along with examples as needed.

Author: I can be more clear here, for sure. The idea is to have YAML as the config format, while all configuration should be _backed_ by Go structs. I can certainly make that more clear

>  - Standard location for all configuration for tooling provided by devbase, et. al

Author: Gr, I'm mad about that. George had me remove that since it was leaking context about internal usage @ Outreach. I agree with you though, I'll workshop this

 - Documentation on how to write Magefile targets
 - Ensure that all targets are "plug-and-play" compatible (e.g. usable by themselves), they shouldn't require stencil modules to be used, and if they do require certain options to work "out of the box" they should be sane defaults and well documented (as per the above goals)

>  - Not breaking, while we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal

Reviewer: Comma splice. Also, could be slightly clearer:
```suggestion
 - Not breaking (backwards-compatible). While we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal.
```

### Non-goals

>  - Expose new functionality, this would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.

Reviewer: Comma splice:
```suggestion
 - Expose new functionality. This would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.
```

Author: You've probably realized by now I am terrible about following the rules of comma usage and "to" vs "too"

>  - Migrating `Makefile` to `Magefile`, while it'd be nice to migrate all of them over, that's also an amount of work to do. While moving to `Magefile` would be good, we shouldn't try to move everything to Go at the same time (TL;DR: Wrap shell were needed).

Reviewer: Comma splice and typo:
```suggestion
 - Migrating `Makefile` to `Magefile`. While it'd be nice to migrate all of them over, that's also an amount of work to do. While moving to `Magefile` would be good, we shouldn't try to move everything to Go at the same time (TL;DR: Wrap shell where needed).
```

## Design Details

### Repository Configuration

> Configuration would live in the root of repository consuming `devbase`, in `.devbase`.

Reviewer: Use consistent voice, here and line 54. Could also use clarity:
```suggestion
Configuration will live in the root of repository consuming `devbase`, in the `.devbase` folder.
```

> Each configuration file would match the relevant `make <target>` target, e.g.

Reviewer: ```suggestion
Each configuration file will match the relevant `make <target>` target with a `.yaml` suffix added, e.g.
```

```bash
$ ls .devbase
build.yaml
e2e.yaml
docker.yaml
```

> **Note**: Configuration should be able to be shared between targets (e.g. `docker-build` and `docker-push` today largely share `docker.yaml`), but this should be limited to only being able to consume the same configuration struct, not two different configs in one. The idea is to allow _closely_ related targets to consume the same configuration only. Complicated things like target a and target b sharing two different configs in a file is expressly what this system is _not_ trying to support.

Reviewer: Brevity / clarity:
```suggestion
**Note**: Configuration should be able to be shared between targets (e.g. `docker-build` and `docker-push` today largely share `docker.yaml`), but this should be limited to only being able to consume the same configuration struct, not having two different configurations live in one file. The idea is to allow _closely_ related targets to consume the same configuration.
```

### Magefile Targets Implementation Details

> Each magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`. The reasoning for a package per target is to make it easier to test these in isolation, as well as (generated) documentation to be localized to the packages (targets) themselves. A library will be provided at `pkg/targets`, which will provide functionality to read configuration and other shared functions between targets.

Reviewer: Be consistent with casing?
```suggestion
Each Magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`. The reasoning for a package per target is to make it easier to test these in isolation, as well as (generated) documentation to be localized to the packages (targets) themselves. A library will be provided at `pkg/targets`, which will provide functionality to read configuration and other shared functions between targets.
```

> ### Magefile Targets Available Day One

Reviewer: It's not clear why you use "day one" here given that this project isn't split into phases. Drop this description?

The following targets will be available/configurable day one:

 - `build`
 - `debug`
 - `release`
 - `test` (includes: `coverage` as `--coverage`, `benchmark` as `--bench`)
 - `lint`
 - `fmt`
 - `e2e`

>  - `docker` (includes: `build`, `publish`)

Reviewer: Does this mean "docker build" and "docker publish" are separate targets?

Just enumerate them instead?

> The rest will still be available for compatibility, but unconfigurable, e.g. `gobuild` will still build go but will not work for non-go repos.

Reviewer: Clarity suggestion:
```suggestion
The other existing targets will still be available for compatibility, but will not be configurable, e.g. `gobuild` will still build go but will not work for non-go repos.
```

See [Configuration Examples](#configuration-examples) for an idea of what options will be available/an idea of some of the steps.

### Testing Framework

> Testing would be done via Go Testing, enabling us to easily write unit tests. Integration tests would, unforuantely, still need to be figured out as they rely on being ran in macOS/Linux in a predictable fashion

Reviewer: ```suggestion
Testing would be done via Go Testing, enabling us to easily write unit tests. Integration tests would, unfortunately, still need to be figured out as they rely on being run in macOS/Linux in a predictable fashion
```

### Configuration Examples

> Below are example configuration objects that would be exposed, note this is not all-inclusive and is purposefully not fully spec'd out to allow for changes during implementation.

Reviewer: ```suggestion
Below are example configuration objects that would be exposed. Note this is not all-inclusive and is purposefully not fully spec'd out to allow for changes during implementation.
```
